CIPs/0008.md

```diff
@@ -0,0 +1,76 @@
+}
+```
+
+Alternatively, this could be set directly on the `Accounts` smart contract directly, causing off-chain storage to effectively replace the existing metadata functionality. This would have the benefit of not duplicating similar functionality.
+
+Either way, this http address should be referred to as the root. The expectation here is that data can accessed via `root/${dataAddress}`. For private data, this should be encrypted to the recipient. Here are some common ways of modeling applications into this format:
+
+- For transaction metadata, the account can encrypt the metadata and put it in a file with the transaction hash as the metadata. Further CIPs could specify a standard for thee structure of that metadata.
+    - `dataAddress`: `transactionHash`
+    - file contents: Encrypted, structured transaction metadata (comments, or otherwise)
+- Data in relation to other accounts could be communicated to other accounts with:
+    - `dataAddress`: `otherAccountAddress`
+    - file contents: Structured data such as:
+        - `sessionKeys`: The session key(s) encrypted with dataEncryptionKey of `otherAccount` on `Accounts.sol`
+        - `attestations`: These could be signed data attestating things such an audit or badges. An attested account could also store the attestations from the attestor here, as a client can just verify the signature of the attestor.
+- Profile information could be stored on this system with:
+    - `dataAddress`: `profileName` or `profilePicture`
+    - file contents: the data unencrypted, or encrypted to session keys available to the recipients
+- Releases:
+    - `dataAddress`: `hashOfRelease`
+    - file contents: The signature of the key on the hash of the release
+
+If no assocation between to the `dataAddress` should be leaked, the `dataAddress` itself could be encrypted. The recipient would then just have to decrypt all `dataAddress`s (which could be available in an index file).
+
+All file contents should always be signed by an authorized signer of the account to prevent tampering by the file hoster.
+
+## Alternative Solutions
+
+I considered doing storing all data on IPFS, however the tooling around IPFS as well as IPNS does not seem user-friendly enough to make a substantial number of users use this directly (or centralize anyways). A generic "HTTP-based file system" is proposed as it is the lowest common denominator. It can be served by a simple HTTP server, even popular VCS static page hosting or specialized service providers. The lock-in effect is minimal since data can be easily migrated, and multiple copies could even be hosted with multiple roots. Thus availabillity concerns relative to on-chain storage can be mitigated relatively well.
+
+## Risks
+
+More reliance on off-chain data and misunderstanding such reliance could lead to inconsistency or poor usability. Ultimately this is traded off against cost of storing data on-chain or in centralized services.
```

Comment:
A related issue is the ability to do a DoS attack on a single host.  One approach is to leave this to the hosting service; another is to provide mirrors by making the root address a list rather than a singleton (but then leading to possible inconsistencies); perhaps another (but less scalable) is to have an index file that is a signed mapping of paths to content hash, then have the directory structure contain content hashes. This opens the door for anyone to host or cache them in the future.

Reply:
I believe I addressed that in the section above, but I do expect a nice trade-off between these concerns to have multiple storage providers with the data mirrored

Reply:
so is `address` going to be an array then?

Reply:
I'd imagine one could have multiple storage claims